File: src/main/kotlin/photon/lib/PushbackStringReader.kt
package photon.lib

import java.io.FilterReader
import java.io.Reader
import java.nio.CharBuffer

class PushbackStringReader(val reader: Reader) : FilterReader(reader) {
  var pushbackValue: Int? = null

  override fun read(): Int {
    synchronized(lock) {
      val c = pushbackValue

      if (c != null) {
        pushbackValue = null
        return c
      } else {
        return super.read()
      }
    }
  }

  fun unread(c: Int) {
    pushbackValue = c
  }

  override fun read(cbuf: CharArray): Int = throw UnsupportedOperationException()
  override fun read(cbuf: CharArray, off: Int, len: Int): Int = throw UnsupportedOperationException()
  override fun read(target: CharBuffer): Int = throw UnsupportedOperationException()
}

File: photon-scala1/src/main/scala/photonscala/lib/PushbackStringReader.scala
package photonscala.lib

import java.io.{FilterReader, Reader}
import java.nio.CharBuffer

class PushbackStringReader(reader: Reader) extends FilterReader(reader) {
  var pushbackValue: Option[Int] = None

  override def read(): Int = {
    lock.synchronized {
      pushbackValue match {
        case Some(c) =>
          pushbackValue = None
          c
        case None => super.read()
      }
    }
  }

  def unread(c: Int): Unit = {
    pushbackValue = Some(c)
  }

  override def read(cbuf: Array[Char]): Int = throw new UnsupportedOperationException
  override def read(cbuf: Array[Char], off: Int, len: Int): Int = throw new UnsupportedOperationException
  override def read(target: CharBuffer): Int = throw new UnsupportedOperationException
}
